cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/AbstractNotificationSubscriberService.java

```diff
@@ -0,0 +1,276 @@
+    long processNotifications() {
+      try {
+        // Gather a local copy of fetched notifications to batch process in one transaction
+        final List<Notification> fetchedNotifications = new ArrayList<>();
+        String lastFetchedMessageId = null;
+        try (CloseableIterator<Message> iterator = messagingContext.getMessageFetcher()
+                                                                   .fetch(NamespaceId.SYSTEM.getNamespace(), topic,
+                                                                          100, messageId)) {
+          LOG.trace("Fetch with messageId = {}", messageId);
+          while (iterator.hasNext() && !stopping) {
+            Message message = iterator.next();
+            lastFetchedMessageId = message.getId();
+            try {
+              Notification notification = GSON.fromJson(new String(message.getPayload(), StandardCharsets.UTF_8),
+                                                        Notification.class);
+              fetchedNotifications.add(notification);
+            } catch (JsonSyntaxException e) {
+              LOG.warn("Failed to decode message with id {}. Skipped. ", message.getId(), e);
+              // lastFetchedMessageId was updated at the beginning of the loop to skip this message in next fetch
+            }
+          }
+        }
+
+        // Sleep for configured number of milliseconds if there are no notifications
+        if (lastFetchedMessageId == null) {
+          return cConf.getLong(Constants.Scheduler.EVENT_POLL_DELAY_MILLIS);
+        }
+
+        final String lastMessageId = lastFetchedMessageId;
+        // Execute processed notifications in one transaction
+        Retries.supplyWithRetries(new Supplier<Void>() {
+          @Override
+          public Void get() {
+            Transactions.executeUnchecked(transactional, new TxCallable<Void>() {
+              @Override
+              public Void call(DatasetContext context) throws Exception {
+                processFetchedNotifications(context, fetchedNotifications);
+                updateMessageId(context, lastMessageId);
+                return null;
+              }
+            });
+            return null;
+          }
+        }, co.cask.cdap.common.service.RetryStrategies.fixDelay(Constants.Retry.RUN_RECORD_UPDATE_RETRY_DELAY_SECS,
+                                                                TimeUnit.SECONDS));
+
+        // Transaction was successful, so reset the failure count to 0
+        failureCount = 0;
+        // The last fetched message id was persisted successfully, update the local field as well.
+        messageId = lastMessageId;
+        return 0L;
+
+      } catch (ServiceUnavailableException e) {
+        SAMPLING_LOG.warn("Failed to contact service {}. Will retry in next run.", e.getServiceName(), e);
+      } catch (TopicNotFoundException e) {
+        SAMPLING_LOG.warn("Failed to fetch from TMS. Will retry in next run.", e);
+      } catch (Exception e) {
```

Comment:
I think it is okay - we need to retry fetching notifications anyway so I think it's better to log it and not continue throwing an error.